interfaces/builtin/unity8.go

```diff
@@ -0,0 +1,190 @@
+       member=DispatchURL
+       peer=(name=com.canonical.URLDispatcher,label=###SLOT_SECURITY_TAGS###),
+
+  # This is needed when the app is already running and needs to be passed in
+  # a URL to open. This is most often used with content-hub providers and
+  # url-dispatcher, but is actually supported by Qt generally.
+  dbus (receive)
+       bus=session
+       path=/###PLUG_DBUS_APPIDS###
+       interface=org.freedesktop.Application
+       member=Open
+       peer=(label=###SLOT_SECURITY_TAGS###),
+
+  # Unity launcher (e.g. app counter, progress, alert)
+  dbus (receive, send)
+       bus=session
+       path=/com/canonical/Unity/Launcher/###PLUG_DBUS_APPIDS###
+       peer=(name=com.canonical.Unity.Launcher,label=###SLOT_SECURITY_TAGS###),
+
+  # Content Hub (pasteboard, file transfers)
+  dbus (receive, send)
+       bus=session
+       interface=com.ubuntu.content.dbus.Service
+       path=/
+       peer=(name=com.ubuntu.content.dbus.Service,label=###SLOT_SECURITY_TAGS###),
+
+  # Lttng tracing is very noisy and should not be allowed by confined apps.
+  # Can safely deny. LP: #1260491
+  deny /{,var/}{dev,run}/shm/lttng-ust-* r,
+`)
+
+var unity8ConnectedPlugSecComp = []byte(`
+recvfrom
+recvmsg
+sendmsg
+sendto
+shutdown
+`)
+
+type Unity8Interface struct{}
+
+func (iface *Unity8Interface) Name() string {
+	return "unity8"
+}
+
+func (iface *Unity8Interface) String() string {
+	return iface.Name()
+}
+
+func (iface *Unity8Interface) dbusAppId(app *snap.AppInfo) []byte {
+	var retval bytes.Buffer
+
+	appidbits := []string{app.Snap.Name(), app.Name, app.Snap.Revision.String()}
+	appid := strings.Join(appidbits, "_")
+
+	for _, value := range appid {
+		if unicode.In(rune(value), unicode.Letter, unicode.Digit) {
+			retval.WriteRune(rune(value))
+		} else {
+			retval.WriteString(fmt.Sprintf("_%2.2x", value))
+		}
+	}
+
+	return retval.Bytes()
+}
+
+func (iface *Unity8Interface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+func (iface *Unity8Interface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		oldTags := []byte("###SLOT_SECURITY_TAGS###")
+		newTags := slotAppLabelExpr(slot)
+		snippet := bytes.Replace(unity8ConnectedPlugAppArmor, oldTags, newTags, -1)
+
+		appidsOld := []byte("###PLUG_DBUS_APPIDS###")
+		var appidsNew bytes.Buffer
+		var appidsList []string
+		for _, app := range plug.Apps {
+			appidsList = append(appidsList, string(iface.dbusAppId(app)))
+		}
+		sort.Strings(appidsList) // makes tests reliable
+		if len(appidsList) == 1 {
+			appidsNew.WriteString(appidsList[0])
+		} else if len(appidsList) > 1 {
+			appidsNew.WriteByte('{')
+			appidsNew.WriteString(strings.Join(appidsList, ","))
+			appidsNew.WriteByte('}')
+		}
+		snippet = bytes.Replace(snippet, appidsOld, appidsNew.Bytes(), -1)
```

Comment:
I strongly feel UAL should be moved to snap.name.command for the AppID. Snappy is a lot different than click. What they share is that on click name_command_revision was *the* AppID everywhere and on snappy snap.name.command is *the* AppID everywhere. The current snapd approach was very actively chosen not to use revision as part of the AppID because there is only every supposed to be one ID for the snap's commands at any given time. Remember, revision changes with local installs without snap assertion (ie, --dangerous), 'snap try', 'snap revert', 'snap refresh', etc. Trying to keep all that straight to map a revision in UAL is adding a lot of complexity. Finally, if you are using name_command_revision as the AppID, the developer has to always keep that context in mind-- that would be meaningless in most security policy (and therefore interfaces). Trusted helpers are no longer able to use libapparmor to get the label of the connecting process and use it as the AppID-- instead they will all have to be modified to maintain a translation.

Is it a requirement that UAL use name_command_revision or is it that it just hasn't happened yet? If it just hasn't happened yet, please do it. If it is a hard requirement, I strongly feel that @niemeyer, @vosst and other stakeholders should work through this now and come up with a proper redesign so that we aren't pulling in tech debt.